detector_test.go

```diff
@@ -1,108 +1,390 @@
+					{
+						Group: []buildpack.GroupBuildpack{
+							{ID: "A", Version: "v1"},
+							{ID: "F", Version: "v1"},
+							{ID: "B", Version: "v1"},
+						},
+					},
+				},
+			})
 
-			_, _, err := lifecycle.BuildpackOrder{
-				{Group: []lifecycle.GroupBuildpack{{ID: "E", Version: "v1"}}},
-			}.Detect(config)
-			if err, ok := err.(*lifecycle.Error); !ok || err.Type != lifecycle.ErrTypeFailedDetection {
-				t.Fatalf("Unexpected error:\n%s\n", err)
+			buildpackStore.EXPECT().Lookup("A", "v1").Return(bpA1, nil)
+			bpA1.EXPECT().ConfigFile().Return(&buildpack.Descriptor{API: "0.3"})
+			bpA1.EXPECT().Detect(gomock.Any())
+
+			buildpackStore.EXPECT().Lookup("F", "v1").Return(bpF1, nil)
+			bpF1.EXPECT().ConfigFile().Return(&buildpack.Descriptor{
+				API: "0.2",
+				Order: []buildpack.Group{
+					{Group: []buildpack.GroupBuildpack{
+						{ID: "C", Version: "v1"},
+					}},
+					{Group: []buildpack.GroupBuildpack{
+						{ID: "G", Version: "v1", Optional: true},
+					}},
+					{Group: []buildpack.GroupBuildpack{
+						{ID: "D", Version: "v1"},
+					}},
+				},
+			})
+
+			buildpackStore.EXPECT().Lookup("C", "v1").Return(bpC1, nil)
+			bpC1.EXPECT().ConfigFile().Return(&buildpack.Descriptor{API: "0.2"})
+			bpC1.EXPECT().Detect(gomock.Any())
+
+			buildpackStore.EXPECT().Lookup("B", "v1").Return(bpB1, nil)
+			bpB1.EXPECT().ConfigFile().Return(&buildpack.Descriptor{API: "0.2"})
+			bpB1.EXPECT().Detect(gomock.Any())
+
+			firstGroup := []buildpack.GroupBuildpack{
+				{ID: "A", Version: "v1", API: "0.3"},
+				{ID: "C", Version: "v1", API: "0.2"},
+				{ID: "B", Version: "v1", API: "0.2"},
 			}
+			firstResolve := resolver.EXPECT().Resolve(
+				firstGroup,
+				detector.Runs,
+			).Return(
+				[]buildpack.GroupBuildpack{},
+				[]lifecycle.BuildPlanEntry{},
+				lifecycle.ErrFailedDetection,
+			)
+
+			buildpackStore.EXPECT().Lookup("G", "v1").Return(bpG1, nil)
+			bpG1.EXPECT().ConfigFile().Return(&buildpack.Descriptor{
+				API: "0.2",
+				Order: []buildpack.Group{
+					{
+						Group: []buildpack.GroupBuildpack{
+							{ID: "A", Version: "v2"},
+							{ID: "B", Version: "v2"},
+						},
+					},
+					{
+						Group: []buildpack.GroupBuildpack{
+							{ID: "C", Version: "v2"},
+							{ID: "D", Version: "v2"},
+						},
+					},
+				},
+			})
 
-			if s := cmp.Diff("\n"+allLogs(logHandler), outputFailureEv1); s != "" {
-				t.Fatalf("Unexpected log:\n%s\n", s)
+			buildpackStore.EXPECT().Lookup("B", "v2").Return(bpB2, nil)
+			bpB2.EXPECT().ConfigFile().Return(&buildpack.Descriptor{API: "0.2"})
+			bpB2.EXPECT().Detect(gomock.Any())
+
+			secondGroup := []buildpack.GroupBuildpack{
+				{ID: "A", Version: "v1", API: "0.3"},
+				{ID: "B", Version: "v2", API: "0.2"},
+			}
+			secondResolve := resolver.EXPECT().Resolve(
+				secondGroup,
+				detector.Runs,
+			).Return(
+				[]buildpack.GroupBuildpack{},
+				[]lifecycle.BuildPlanEntry{},
+				lifecycle.ErrFailedDetection,
+			).After(firstResolve)
```

Comment:
Since each buildpack toml `Detect` happens in a go func, we can't use `gomock.InOrder`